File: algorithms/graham.py
import math
from functools import cmp_to_key
from collections import namedtuple


P0 = ()


def get_distance(get_pivot, point):
    return math.hypot(point.x - get_pivot.x, point.y - get_pivot.y)


def get_determinant(P, A, B):
    return (P.x * A.y + P.y * B.x + A.x * B.y) -\
           (B.x * A.y + B.y * P.x + A.x * P.y)


def get_pivot(P):
    i = 0
    while i < len(P):
        if P[i].y < P[0].y or (P[i].y == P[0].y and P[i].x > P[0].x):
            P[i], P[0] = P[0], P[i]
        i += 1
    return P[0]


def sort_by_angle(A, B):
    # pontos colineares: escolhe-se o mais próximo do pivô
    # se o get_determinante for 0, os pontos estão alinhados
    if get_determinant(P0, A, B) == 0:
        if get_distance(P0, A) < get_distance(P0, B):
            return -1
        if get_distance(P0, A) > get_distance(P0, B):
            return 1
        return 0

    alfa = math.atan((A.y - P0.y)/(A.x - P0.x))
    beta = math.atan((B.y - P0.y)/(B.x - P0.x))

    if alfa < beta:
        return -1
    if alfa > beta:
        return 1
    return 0


def make_convex_hull(P):
    N = len(P)

    # Corner case: com 3 vértices ou menos, P é o próprio convex hull
    if N <= 3:
        return P

    global P0
    P0 = get_pivot(P)
    P.remove(P0)

    P.sort(key=cmp_to_key(sort_by_angle))

    N = len(P)

    # o primeiro ponto é igual ao último
    ch = [P[N-1], P0, P[0]]

    i = 1
    while i < N:
        j = len(ch) - 1

        # se o get_determinante for positivo, a orientação de manteve
        if get_determinant(ch[j-1], ch[j], P[i]) > 0:
            ch.append(P[i])
            i += 1
        else:
            ch.pop()

    return ch


Point = namedtuple('Point', 'x y')

P = [
    Point(0, 0),
    Point(5, 3),
    Point(8, -2),
    Point(4, 4),
    Point(2, 1),
    Point(2, 5),
    Point(3, -1),
    Point(7, 2),
    Point(5, 0),
    Point(0, 4),
    Point(1, -1),
    Point(7, -2),
    Point(6, 4),
    Point(6, 0),
    Point(1, 3)
]

ch = make_convex_hull(P)
i = 0
for i in range(len(ch)):
    print(f"{i+1}: ({ch[i].x}, {ch[i].y})")

File: src/convex_hull/graham.cpp
/* 
Adaptação do código de Edson Alves, disponível em:
https://github.com/edsomjr/TEP/blob/master/Geometria_Computacional/slides/CH-1/graham.cpp
*/ 

#include <bits/stdc++.h>
#include <chrono>

using namespace std;
using ll = long long int;

template<typename T> struct Point{
    T x, y;

    double get_distance(const Point& P) const{
        return hypot(x - P.x, y - P.y);
    }
};

template<typename T> T get_determinant(const Point<T>& P, const Point<T>& A, const Point<T>& B){
    return (P.x * A.y + P.y * B.x + A.x * B.y) -
           (B.x * A.y + B.y * P.x + A.x * P.y);
}

template<typename T> Point<T> get_pivot(vector<Point<T>>& P){
    for (size_t i = 1; i < P.size(); ++i)
        if (P[i].y < P[0].y or 
            (P[i].y == P[0].y and P[i].x > P[0].x))
                swap(P[0], P[i]);

    return P[0];
}

template<typename T> void sort_by_angle(vector<Point<T>>& P){
    auto P0 = get_pivot(P);

    sort(P.begin() + 1, P.end(),
        [&](const Point<T>& A, const Point<T>& B){
            // pontos colineares: escolhe-se o mais próximo do pivô
            if (get_determinant(P0, A, B) == 0) // se o get_determinante for 0, os pontos estão alinhados
                return A.get_distance(P0) < B.get_distance(P0);

            auto alfa = atan2(A.y - P0.y, A.x - P0.x);
            auto beta = atan2(B.y - P0.y, B.x - P0.x);

            return alfa < beta;
        }
    );
}

template<typename T> vector<Point<T>> make_convex_hull(const vector<Point<T>>& points){
    vector<Point<T>> P(points);
    auto N = P.size();

    // Corner case: com 3 vértices ou menos, P é o próprio convex hull
    if (N <= 3)
        return P;

    sort_by_angle(P);

    vector<Point<T>> ch;
    ch.push_back(P[N - 1]); // o primeiro ponto é igual ao último
    ch.push_back(P[0]);
    ch.push_back(P[1]);

    size_t i = 2;

    while (i < N) {
        auto j = ch.size() - 1;

        // se o get_determinante for positivo, a orientação se manteve
        if (get_determinant(ch[j - 1], ch[j], P[i]) > 0)
            ch.push_back(P[i++]);
        else
            ch.pop_back();
    }

    return ch;
}

int main(){
    vector<Point<ll>> P;
    ll x, y = 0;

    while(cin >> x >> y){
        P.push_back({x, y});
    }

    auto begin = std::chrono::high_resolution_clock::now();

    auto ch = make_convex_hull(P);

    auto end = std::chrono::high_resolution_clock::now();
    auto elapsed = std::chrono::duration_cast<std::chrono::microseconds>(end - begin);

    fprintf(stderr, "Time measured: %f seconds.\n", elapsed.count() * 1e-6);


    for (size_t i = 0; i < ch.size(); ++i)
        cout << ch[i].x << " " << ch[i].y << endl;

    return 0;
}
